fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1676,6 +1678,34 @@ private static void checkStoreHeaderInternal(@Nonnull RecordMetaDataProto.DataSt
         }
     }
 
+    private void beginRecordStoreStateRead() {
+        recordStoreState.updateAndGet(state -> {
```

Comment:
Given that these four methods only update the state of the `recordStoreState`, I'm not sure the `updateAndGet` call makes sense. It seems like, at most, they should read the `recordStoreState` from the reference and then call begin/end read/write. (They could also validate that `RecordStoreState` is non-null.)

I could also see the argument that callers should extract the record store state from the reference and then call begin/end read/write directly on the (dereferenced) `recordStoreState`.

Reply:
Oh, I think I have a better idea of what this is trying to do now that I've thought more on it--the idea is that if someone updates the record store state multiple times, this function (and its implicit retry loop at the atomic reference level) will retry the update on the new record store state, so the operation always does what's expected (i.e., update the "in flight reads" value while leaving the record store state otherwise in tact).

See, my thought was that the same effect would have been arranged by always ensuring that `.set` is called at most once on the given record store state ref (which the possible exception of the section that saves a copy with all new indexes in a write-only state temporarily). Hm, but maybe this is more resilient. I'm not sure.

Reply:
I think so. It probably needs a comment explaining why this is done inside an “update” lambda even though all it does is modify internal state.

Reply:
sure.